File: Difficulty: Hard/Max Circular Subarray Sum/max-circular-subarray-sum.cpp
class Solution {
  public:
    int maxCircularSum(vector<int> &arr) {
        int n = arr.size();
        int total = 0;
        int minSum = INT_MAX, minEnding = 0;
        int maxSum = INT_MIN, maxEnding = 0;
        
        for(int i=0; i<n; i++) {
            total += arr[i];
            
            minEnding = min(minEnding + arr[i], arr[i]);
            minSum = min(minSum, minEnding);
            
            maxEnding = max(maxEnding + arr[i], arr[i]);
            maxSum = max(maxSum, maxEnding);
        }
        
        if(total == minSum) return maxSum;
        
        return max(maxSum, total - minSum);
    }
};

File: Difficulty: Hard/Max Circular Subarray Sum/max-circular-subarray-sum.java
class Solution {
    public int maxCircularSum(int arr[]) {
        int currMax = arr[0], maxSum = arr[0];
        int currMin = arr[0], minSum = arr[0];
        int totalSum = arr[0];
        int n = arr.length;
        for(int i=1; i<n; i++) {
            currMax = Math.max(arr[i], currMax + arr[i]);
            maxSum = Math.max(maxSum, currMax);
            
            currMin = Math.min(arr[i],currMin + arr[i]);
            minSum = Math.min(minSum, currMin);
            
            totalSum += arr[i];
        }
        
        if(totalSum == minSum)
            return maxSum;
            
        return Math.max(maxSum, totalSum - minSum);
    }
}
